### DISCUSSION TITLE
Coupling Urban Canopy Model as a subLSM for NoahMP.4.0.1

### DISCUSSION THREAD

**Discussion by jyoo8** [NASA-LIS/LISF#1329] [Q&A]
Hi,
I have prepared Urban Canopy Model as a subLSM for NoahMP.4.0.1 as Crocus model and Snowmodel in the LIS.
I think I have set up everything for the coupling but I am getting a Segmentation fault error as soon as I execute the LIS retrospective run.

It seems that the error occurs when the LIS runs through LIS_sublsm_FTable.c file.
Specifically, at the function of void FTN(sublsminit)(char *j,int *kk, int len).

The section of the code in the LIS_sublsm_FTable.c file looks as below:

//BOP
// !ROUTINE: sublsminit
// \label{sublsminit}
//
// !INTERFACE:
void FTN(sublsminit)(char *j,int *kk, int len)
//
//  
// !DESCRIPTION: 
//  Invokes the routine from the registry to perform
//  land surface model initialization
// 
//  \begin{description}
//  \item[j]
//   name of the SUBLSM
//  \item[kk] 
//   index of the SUBLSM
//  \end{description}
//EOP
{ 

  struct sublsminitnode* current;
  int found ; 

  current = sublsminit_table;

  while(strcmp(current->name,j)!=0){
    current = current->next;

    if(current==NULL) {
      printf("****************Error****************************\n"); 
      printf("init routine for SUBLSM %s is not defined\n",j); 
      printf("program will seg fault.....\n"); 
      printf("****************Error****************************\n"); 
    }
  }
  current->func(kk); 

}

I noticed that "current" in my case is "null" even after  the line "current = sublsminit_table;" and the LIS crashed even before going into the  while(strcmp(current->name,j)!=0){} section.

I think this is related with pnode in the void FTN(registersublsminit)(char *j, void (*func)(int*),int len) but my LIS retrospective run didn't go through the void FTN(registersublsminit)(char *j, void (*func)(int*),int len) section. 

Also, I do have LIS_sublsm_pluginMod.F90 for Urban (Canopy Model) under plugins directory as below:

#if ( defined SM_Urban )
   call registersublsminit(trim(LIS_UrbanId)//char(0),Urban_ini)
   call registersublsmsetup(trim(LIS_UrbanId)//char(0),Urban_setup)
   call registersublsmf2t(trim(LIS_UrbanId)//"+"//&
        trim(LIS_retroId)//char(0),Urban_f2t)
   call registersublsmf2t(trim(LIS_UrbanId)//"+"//&
        trim(LIS_agrmetrunId)//char(0),Urban_f2t)
   call registersublsmrun(trim(LIS_UrbanId)//char(0),Urban_main)
   call registersublsmrestart(trim(LIS_UrbanId)//char(0),Urban_readrst)
   call registersublsmdynsetup(trim(LIS_UrbanId)//char(0),Urban_dynsetup)
   call registersublsmwrst(trim(LIS_UrbanId)//char(0),Urban_writerst)
   call registersublsmfinalize(trim(LIS_UrbanId)//char(0),Urban_finalize)

   !wirings between NoahMP and URBAN
   call registersublsmsetLSMimport(trim(LIS_UrbanId)//char(0),&
        Urban_setLSMimport)

#if ( defined SM_NOAHMP_4_0_1 )
   call registerlsm2sublsmgetexport(trim(LIS_noahmp401Id)//"+"//&
        trim(LIS_UrbanId)//char(0),NoahMP401_getURBANexport)
   call registerlsmsetsublsmimport(trim(LIS_noahmp401Id)//char(0),&
        NoahMP401_setURBANimport)
   call registersublsm2lsmgetexport(trim(LIS_noahmp401Id)//"+"//&
        trim(LIS_UrbanId)//char(0),Urban_getLSMexport)
#endif
#endif

Any help and comment will be highly appreciated.
Thank you.

**Comment by dmocko** [comment]
Hi @jyoo8 

Quick question - did you add the proper lines in your lis/make/default.cfg file to add this subLSM?

**Reply by jyoo8** [reply]
Hi @dmocko,

Here is the modification that I added in the lis/make/default.cfg file


[Urban]
enabled: False
macro: SM_Urban
path: surfacemodels/land/subLSM/Urban,
      surfacemodels/land/subLSM/Urban/phys,
      surfacemodels/land/subLSM/Urban/cplLSM

Thank you!

**Comment by dmocko** [comment]
That "enabled" should be set to "True" - or your LISF/lis/make/user.cfg has to have a line: `"Urban: On"`.

**Reply by jyoo8** [reply]
Hi @dmocko, 
I made changes in the file. But still I don't get the dependency files generated. Do you know where I should make changes too? 
Thank you.

**Comment by dmocko** [comment]
I assume that you ran "gmake realclean" in your LISF/lis/make directory, and then re-compiled?

**Reply by jyoo8** [reply]
I am compiling the LISF within NU-WRF using the compilation command, "./build.sh -o debug,cleanfirst  lis" which should be equivalent to the "gmake realclean", I guess. Thank you.

**Reply by jyoo8** [reply]
Hi @dmocko, 
Actually, "make realclean" was necessary before I continued to compile. Thank you.

Also, I made a few change in the default.cfg file for my Urban subLSM model as below:
[Urban]
enabled: True
macro: SM_Urban
path: surfacemodels/land/subLSM/urban,
      surfacemodels/land/subLSM/urban/cplLSM
dependent_comps: NoahMP.4.0.1
NoahMP.4.0.1 path: surfacemodels/land/noahmp.4.0.1,
                   surfacemodels/land/noahmp.4.0.1/phys

Here in particular, I added lines for "dependent_comps:" and "NoahMP.4.0.1 path:".
I did it because the Urban Canopy Model uses Urban related module_sf_ xxxxxxxx.F90 files that exist under surfacemodels/land/noahmp.4.0.1/phys directory.

I am not sure if these lines are correct to implement as such.
But dependency files for Urban have been generated now.

Regardless, when I compile the LIS again I got error that looks like highly relevant as below:

error: can't open file module_sf_noahmpdrv_401.o for write
compilation aborted for ../surfacemodels/land/noahmp.4.0.1/phys/module_sf_noahmpdrv_401.F90 (code 1)
Makefile:151: recipe for target 'module_sf_noahmpdrv_401.o' failed
make: *** [module_sf_noahmpdrv_401.o] Error 1
make: *** Waiting for unfinished jobs....


Any comments will be highly appreciated.
Thank you!


Currently I am not sure which

**Reply by jvgeiger** [reply]
Hello @jyoo8,

These lines

```
dependent_comps: NoahMP.4.0.1
NoahMP.4.0.1 path: surfacemodels/land/noahmp.4.0.1,
surfacemodels/land/noahmp.4.0.1/phys
```
are not needed in your default.cfg file.  `dependent_comps` is used for setting features like irrigation, DA, or OPTUE.  The dependency generator will find NoahMP 4.0.1.

Have you tried compiling just LIS - meaning not compiling LIS as part of NU-WRF?

Thanks,
Jim

**Reply by jyoo8** [reply]
Hi @jvgeiger,
Thank you very much for your notes!
I have not tried compiling just LIS yet but I can do that too.
Thank you.

**Reply by jyoo8** [reply]
Hi @jvgeiger,
I removed and compiled it again.
Now I am getting error with noahmp.4.0.1/noahmp_driver_401.F90 which has not been modified at all for the Urban model.
Part of error messages looks as shown below:
../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90(768): error #6633: The type of the actual argument differs from the type of the dummy argument.   [ISNOWINOUT]
       isnowinout   , tvinout      , tginout      , caniceinout  , canliqinout  , eahinout     , & ! in/out Noah MP only
-------^
../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90(768): error #6633: The type of the actual argument differs from the type of the dummy argument.   [TGINOUT]
       isnowinout   , tvinout      , tginout      , caniceinout  , canliqinout  , eahinout     , & ! in/out Noah MP only
-------------------------------------^
../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90(774): error #6633: The type of the actual argument differs from the type of the dummy argument.   [PGSINOUT]
       pgsinout     ,                                                   & ! in/out Noah MP only
-------^
../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90(777): error #6633: The type of the actual argument differs from the type of the dummy argument.   [T2MVOUT]
       t2mvout , t2mbout , q2mvout , q2mbout , relsmcout,          & ! out Noah MP only
-------^
../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90(786): error #6633: The type of the actual argument differs from the type of the dummy argument.   [PARAMETERS]
       parameters, &
-------^
../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90(787): error #6633: The type of the actual argument differs from the type of the dummy argument.   [FLDSTOIN]
       rivstoin,fldstoin,fldfrcin,                                 &
----------------^
../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90(791): error #6633: The type of the actual argument differs from the type of the dummy argument.   [IDS]
       ids,ide,  jds,jde,  kds,kde,                                &
-------^
../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90(791): error #6633: The type of the actual argument differs from the type of the dummy argument.   [IDE]
       ids,ide,  jds,jde,  kds,kde,                                &
-----------^
../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90(747): error #6631: A non-optional actual argument must be present when invoking a procedure with an explicit interface.   [KTS]
  call noahmplsm_401  (LIS_rc%udef,  & ! in : LIS undefined value (David Mocko)


......


compilation aborted for ../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90 (code 1)
Makefile:151: recipe for target 'noahmp_driver_401.o' failed
make: *** [noahmp_driver_401.o] Error 1
make: *** Waiting for unfinished jobs....

**Comment by jvgeiger** [comment]
Where is your working directory on discover?  I will take a closer look.

**Reply by jyoo8** [reply]
Hi @jvgeiger,
I created a git repo at /discover/nobackup/jyoo8/NU-WRF_dev/urban_cloned .
Thank you very much for your time!

**Comment by jvgeiger** [comment]
Hello @jyoo8,

The code that you gave me was missing the [Urban] section in the make/default.cfg file.  I added it.  Note that you may have a typo.  You specified

```
path: surfacemodels/land/subLSM/Urban,
```

but the path is 

```
surfacemodels/land/subLSM/urban
```

Uppercase U vs lower case u.  I corrected the default.cfg file.

I get this error:

```
surfacemodels/land/subLSM/urban/Urban_driver.F90(853): error #6784: The number of actual arguments cannot be greater than the number of dummy arguments.   [NOAHMP_URBAN]
```

You are calling noahmp_urban with too many variables.

**Reply by jyoo8** [reply]
Hi @jvgeiger,
Thank you very much for your efforts, Jim.
I was able to reproduce the same error now with the newly cloned repository at /discover/nobackup/jyoo8/NU-WRF_dev/urban_cloned/nu-wrf-dev.

Interestingly, the original version of the cloned repository produced the previous error still:
 -I../surfacemodels/land/rdhm.3.5.6/ -I../surfacemodels/land/rdhm.3.5.6/sac/ -I../surfacemodels/land/rdhm.3.5.6/frz/ -I../surfacemodels/land/rdhm.3.5.6/snow17/ -I../surfacemodels/land/rdhm.3.5.6/sachtet/ ../surfacemodels/glacier/noahmp.3.9.1.1/noahmpglacier3911_readcrd.F90
ifort: command line warning #10434: option '-nomixed_str_len_arg' use with underscore is deprecated; use '-nomixed-str-len-arg' instead
../surfacemodels/land/noahmp.4.0.1/noahmp_driver_401.F90(768): error #6633: The type of the actual argument differs from the type of the dummy argument.   [ISNOWINOUT]
       isnowinout   , tvinout      , tginout      , caniceinout  , canliqinout  , eahinout     , & ! in/out Noah MP only
-------^

Also, there was not the section for the Urban in the make/default.cfg in the the newly cloned repository!
Strange.

It seems that the original repository at /discover/nobackup/jyoo8/NU-WRF_dev/nu-wrf_urban202304/nu-wrf-dev may have some mysterious issue, which cannot be resolved easily.

Anyway, let me fix the issue in the Urban_driver.F90 for now.
Thank you very much for your time.
Have a nice weekend!

**Reply by jyoo8** [reply]
Hi @jvgeiger, 
I was able to compile the LIS along with Urban Canopy Model and I am testing the code now.
Currently running on the retrospective mode and it ran through reading in config file of the Urban Canopy Model section successfully.
Now I am getting error with ESMF_StateAdd for an exchange variable, lhField.
Interestingly, there were three other exchange variables before the lhField in which ESMF_FieldCreate and ESMF_StateAdd have been completed successfully. 
I don't know why the lhField didn't pass the section.
Do you have any idea on where I need to make sure?
Thank you very much for your valuable inputs!

**Reply by jyoo8** [reply]
Hi @jvgeiger,
I figured out that special characters (e.g., /, &, etc. ) cannot be used in the name parameters of ESMF_FieldCreate as well as duplications of the names among variables.
I was able to pass the section.
Thanks!

**Comment by jvgeiger** [comment]
Great.
